### ISSUE TITLE
Use dor-services thumb method for determining the thumbnail for display

### ISSUE THREAD

**Issue by LynnMcRae** [sul-dlss/argo#823]
The Rules of Thumb (https://consul.stanford.edu/display/chimera/The+Rules+of+Thumb) have been consolidated into a DOR Services method which can be used by the application when offering a thumbnail on the item detail page.

**Comment by LynnMcRae** [comment]
See DOR-Services method described at sul-dlss/dor-services#236

**Comment by peetucket** [comment]
This is not going to be straightforward.  The image shown for a thumbnail comes from the argo solr document in a field called "first_shelved_image_ss", as shown here: https://github.com/sul-dlss/argo/blob/master/app/helpers/argo_helper.rb#L26

The correct thing to do would be to index the new "thumb" answer into a different field (or switch the field name and index into that new field) and then use this new field.  This requires:

1. A full republish of all images to get the <thumb> element in publicXML.
2. Changing the argo indexer to use this new attribute to populate the new field in solr.
3. Re-indexing everything in argo.
4. Update the argo code to use the new field.

Alternatively, you could use the latest dor-services gem to get the answer, but this requires loading the object from Fedora, and I am assuming the argo item show page must be built purely from solr (instead of fedora) for performance reasons.

**Comment by LynnMcRae** [comment]
While we construct the results page from the solr index, the detail page references Fedora here and there.  But I would be happy to defer this for now given it's sort of a not-so-best effort to begin with (only works for accessioned/shelved items).  I'll make a ticket to add the thumb reference to the index ... seems useful, but who knows when we'll care to get around to it.

**Comment by jmartin-sul** [comment]
IIRC, the search results pages and the front page are built purely from solr results, but the detail page does load the object from fedora.  would have to double-check that to be sure, but my recollection corresponds to lynn's there.

**Comment by peetucket** [comment]
Oh, well, in that case, it would be pretty straightforward -- we just ignore what is in solr when deciding which image to show and the update would probably be to a single helper method.

**Comment by jmartin-sul** [comment]
i would also vote for correcting what's stored in the old solr field, or removing that field entirely if its obsolete.

**Comment by LynnMcRae** [comment]
A retrospective reindexing is not really needed as virtually none but a few test objects have a thumb attribute that would cause the answer to be different, and we know who they are. Any change to the thumb selection going forward would be reindexed when that happened. Can you alias a solr field so the name stays descriptive of the content?

**Comment by LynnMcRae** [comment]
Can this ticket just become a solve-it-by-improving the indexed field, no Argo code changes? As opposed to a separate ticket for that strategy. I would vote for the index approach.

**Comment by jmartin-sul** [comment]
i'm not sure i see the advantage of that... it'd require a dor-services change, and reindexing, right?  also, i hadn't realized earlier what the current solr field name is... to put something else in that field would seem like turning it into a misnamed field, to me.  it may not currently be what we want for a thumb, but it does contain what its name indicates, AFAICT (though one possibly surprising thing is that "first" seems to indicate first listed in the XML, not first shelved chronologically).

anyway, i wasn't proposing a reindexing above -- i was agreeing with what i thought you were suggesting, that we use the new dor-services method in argo since we're already loading the object from fedora.  i'd also proposed possibly switching the solr field to contain what we consider the thumb by the new logic.  i'd retract that considering what i missed about the name, and just stick with the other option mentioned of either leaving that solr field as-is, or discontinuing writing it if nothing references it anymore.

calling the new dor-services method in argo instead of using the old solr field necessitates an argo change, leaves the option for a dor-services change, and needs no reindexing.  existing solr documents with the `first_shelved_image_ss` field wouldn't be doing any harm because the field wouldn't be incorrect, per se, and nothing that cares about thumbs should refer to it anymore anyway.  but changing what's indexed seems to require a dor-services change, and either living with a really misleading field name, or changing argo to refer to a different solr field, plus reindexing.  just modifying argo seems like less work, and no less "right" an approach (unless i'm missing something).

**Comment by LynnMcRae** [comment]
Yes, it would mean a DOR Services change, but not a reindexing, because the answer will vary from the "first image" answer only with new objects going forward.  I was pretty sure SOLR has a field alias capability, but apparently that's only for query, not within the schema itself, so I wish it had been given a functional name from the start. Note that stop populating it rather than simply abandoning it would mean a DOR Services change as well. 

A primary motivation here might be efficiency for an already frustratingly slow app by having the answer already computed. A realtime calculation means adding a weighty DOR fetching/parsing thing that has to happen before we write the page, unless that already is or can be an asynchronous call. If we added a new field because of the naming thing, then Argo could look in both places and still be way faster than running the rules against the DOR object? Or am I over-concerned about the speed of that?

**Comment by andrewjbtw** [comment]
@jmartin-sul do you know if this ticket is still relevant as-is or does it need to be updated and/or should there be a new ticket to reflect current code?

**Comment by jmartin-sul** [comment]
@andrewjbtw 💬 
> @jmartin-sul do you know if this ticket is still relevant as-is or does it need to be updated and/or should there be a new ticket to reflect current code?

haven't looked at this in quite a while, but it does appear from a quick search through the code and a look at a detail page that the mechanism described in this ticket is still in use.  whether the ticket is still actionable/worthwhile, i'm not sure... but the referenced code looks like it's still there.

code:
* https://github.com/sul-dlss/argo/blob/4972727465f270884c89f6f02b370fd7bbdc5f07/app/helpers/argo_helper.rb#L22-L23
* https://github.com/sul-dlss/argo/blob/4972727465f270884c89f6f02b370fd7bbdc5f07/app/helpers/argo_helper.rb#L8-L16 (constructs a IIIF link like the one below)

example object and thumb:
* https://argo.stanford.edu/view/druid:kh677jc8585
  * https://stacks.stanford.edu/image/iiif/kh677jc8585%2Fkh677jc8585_img_1/full/!400,400/0/default.jpg
  * https://argo.stanford.edu/view/druid:kh677jc8585.json (search for `first_shelved_image_ss`)

FWIW, i haven't traced through the detail page rendering code, so maybe the above isn't actually used as i think it is.  i just searched the argo code for `thumb` and went from there.

@jcoyne might have a firmer grasp on the current state of things?

**Comment by jcoyne** [comment]
This is not something we are going to be able to accommodate as Argo no longer uses the dor-services gem.
